server/etcdserver/bootstrap.go

```diff
@@ -322,29 +414,42 @@ func bootstrapWithWAL(cfg config.ServerConfig, st v2store.Store, be backend.Back
 	} else {
 		cfg.Logger.Info("No snapshot found. Recovering WAL from scratch!")
 	}
+	return snapshot, be, nil
+}
 
-	r := &bootstrappedServer{}
-	if !cfg.ForceNewCluster {
-		r.raft = bootstrapRaftFromWal(cfg, snapshot)
-	} else {
-		r.raft = bootstrapRaftFromWalStandalone(cfg, snapshot)
+func (c *bootstrapedCluster) Finalize(cfg config.ServerConfig, s *bootstrappedStorage) error {
+	if !s.wal.haveWAL {
+		c.cl.SetID(c.nodeID, c.cl.ID())
+	}
+	c.cl.SetStore(s.st)
+	c.cl.SetBackend(schema.NewMembershipBackend(cfg.Logger, s.backend.be))
+	if s.wal.haveWAL {
+		c.cl.Recover(api.UpdateCapability)
+		if c.cl.Version() != nil && !c.cl.Version().LessThan(semver.Version{Major: 3}) && !s.backend.beExist {
```

Comment:
This seems like a very complex if statement to me, could we comment or separate out to its own function maybe to just clarify what we are checking here. As the error msg implies only db file is missing.

Reply:
DOne.